Update to touch handling

rschloen opened the pull request:
**Manual control**
Touch handling was changed to use the touch start to send commands as well as touch move, so cases where an actuator button is just clicked without moving the mouse still sends the command. Commands will also be resent at an interval after the touch starts and after moving stops while the mouse is still down. This supports touch and hold (without moving), and holding after moving. 

**Prepared Sequences**
Also added touch support for the preset sequences. For sweep, which can happen in 8 directions, moving cursor in a direction triggers appropriate sequence.  

**Multitouch**
Changed how commands are compiled to support multiple active actuators at once.

rschloen committed:
Used inline style to set initial orientation of actuators

Added setRotationStyle and setScaleStyle to set the overall style, or
transforms, so the actuators are rotated horizontally or vertically for
the different options. The actuators can still be rotated and reset to
their initial position. Scale is currently intended for overlaying an
inactive array of actuators over the option link button, like to mark the
location on the body(WIP).

rschloen committed:
Added images to replace buttons for options

Instead of a button, the background of the div container in the option
link window is set to an image, such as the location of actuators on the
body.

rschloen committed:
Improved device response handling

Errors from the device are handled better, giving the user more feedback in
the message field and preventing actions that should be allowed if an error
is thrown (Ex: showing a device as added when it was not found by antenna)
Message variable was added to the store to make it easier for all components
to access.

Also fixed bug in actuator command. JSON did not like the '00000001'
format so the stings are converted into their decimal forms before being sent

rschloen committed:
Added stichable sized arrays of hexagons

Added drop down menu to switch between stichable and full arrays.

arrayType is a prop so it can also be set when a component is created.

rschloen committed:
Added support for PreparedSequences

Added components to PreparedSequences page.
Added props to set if a hexagon component is for a preset or not. If it is
for a preset, command is sent for that presetName

Checks mouse movement for sweep preset to trigger sweep in swiped direction.

Added OP_mode to store and act_command to send preset commands.

rschloen committed:
Changed handling of touch events

Moved code to create function to send command.
Moved code to create function to find active actuators from mouse pose.

Added checking for active actuators to touch start and set an interval that
will repeat the check to handle holding down an actuator without moving.

Changed touch move to handle moving then stopping to hold (no touch end)
so that the check will repeat at an interval (similar to touch start).

The interval is cleared in touch end.

rschloen committed:
Made style changes in css

Hexagon components no longer have moveable control box visible.

Buttons for selecting automatic/preset options now change color when focused on
Useful to know what option is active.

rschloen committed:
Added visual feedback for presets

When a preset is triggered, the corresponding ui actuators will light up in
the same sequence as the device. Timing is currently hard coded at 500ms.

Hard coded the actuator sequence in the store.

rschloen committed:
Support for multitouch

Changed activeHexagon to an empty array when reset. When checking for
active actuators, add new clicked actuators to activeHexagon.

When building the commands, loop through activeHexagon array and combine
bits (bitwise or with new final variable) to activate multiple actuators.

rschloen committed:
Started to add stichable arrays

Started to added draggable and snappable to Moveable object.

Started to add function to add stichables in Automatic page.

rschloen committed:
Continued work on displaying assets behind arrays

Tried adding another div container for both components in Hexagons (array
and rotation options) to have a background behind them.

background image added a prop for Hexagons so it can be specific to an array.

Tried to fix position of small array on button link in automatic(WIP).

rschloen committed:
Improved handling of touch and multitouch events

Needed to add support to touchEvents, not just mouse events, but both events will
use the same handler since they do pretty much the same thing.

When handling touch start, checks if it is a mouse event or touch event by looking
at length of targetTouches (0 if mouse event). The distinction between event types
is needed since cursor location is calculated differently. If its a touch event,
extra math is needed to find position relative to target container since it is not
available inherently. Everywhere cursor position was used, conditional was added to
change calculation depending on if isTouch or not. target Rectangle bounds found onMount.

Length of targetTouches (numTouches) is used to loop through all touches when finding
active hexagons, and sendCommandBlocks was moved outside of the loops so it would
only send fully complied blocks when there are multiple touches.

Added additional clearInterval/clearTimeout to start and end handlers to prevent
to many intervals from being created or not all intervals being cleared after touch
ends.

Moved X/Y end from handleTouchEnd to handleTouchMove because targetTouches becomes
undefined before the position can be calculated once the touch is ended.

rschloen committed:
Modified scaling of actuators

Use screen size to calculate a scaling factor that gets multiplied by the
hexagonSideLength and baseSpacing to scale the size of the actuator hexagons
based on window size. Also used to make the array overlay for the option
buttons in automatic.

trueb2 commented on src/Utils/Devices.svelte:
```diff
@@ -20,13 +19,19 @@ function handleClickDevice() {
         searching = false;
+        if (result.hasOwnProperty('Failure')){
+            deviceButtonMessage = "Re-attempt Search";
+            $message = result['Failure']['message'];
```
Is it okay for message to be undefined?

trueb2 commented on src/Utils/Devices.svelte:
```diff
@@ -20,13 +19,19 @@ function handleClickDevice() {
+        if (result.hasOwnProperty('Failure')){
+            deviceButtonMessage = "Re-attempt Search";
+            $message = result['Failure']['message'];
+        }else{
+        $message = "Connected to new Device"; 
         $devices = [...$devices, newDevice];
         deviceButtonMessage = "Find";
+        }
```
We might as well format the document.

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -1,16 +1,23 @@
-export let message = 'starting ...';
-
+export let activeHexagon = [];
+export let orientation = "horizontal";
+export let arraySize = "normal";
```
Mathematical arguments will make this much easier to use in other situations. I expect overlaying the hexagons on the smaller options will involve a refactor of these arguments or nesting of components that compute the arguments.

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -21,39 +28,55 @@ let moveable;
+            allActuator |= shiftActuator;
+            binActuator = Math.abs(allActuator).toString(2).padStart(32,'0'); 
+            binActuator = binActuator + zero.repeat(96);
+        }
```
If we have any performance problems, this needs to not use strings.

This approach will only work when multi-touch is within the same actuator block. 
* Instead, a correct approach with this style of accumulation would keep a blockNActuator where we accumulate `blockNActuator |= shiftActuator`. 
* Rather than using the current approach, we might as well
```
// Given actuators = [26, 126]
// Produce [
//     0,0,0,0b00000010,
//     0,0,0,0,
//     0,0,0,0b01000000
// ]
let blockBytes = [...Array(16)].keys().map(i => 0);
for(let actuator of actuators) {
  const block  = Math.floor(actuator / 32)
  const blockByte = Math.floor((actuator % 32) / 8)
  blockBytes[(block * 32) + blockByte] |= ((actuator % 32) - (blockByte * 8))
}
return blockBytes
```

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -21,39 +28,55 @@ let moveable;
-    block32_63.set([b1_0, b1_1, b1_2, b1_3]);
-    block64_95.set([b2_0, b2_1, b2_2, b2_3]);
-    block96_127.set([b3_0, b3_1, b3_2, b3_3]);
+    block0_31.set([parseInt(b0_0,2), parseInt(b0_1,2), parseInt(b0_2,2), parseInt(b0_3,2)]);
+    block32_63.set([parseInt(b1_0,2), parseInt(b1_1,2), parseInt(b1_2,2), parseInt(b1_3,2)]);
+    block64_95.set([parseInt(b2_0,2), parseInt(b2_1,2), parseInt(b2_2,2), parseInt(b2_3,2)]);
+    block96_127.set([parseInt(b3_0,2), parseInt(b3_1,2), parseInt(b3_2,2), parseInt(b3_3,2)]);
```
It would be good to change the getBytesForActuators to use integers instead of strings.

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -148,64 +171,178 @@ $: drawableHexagons = hexagons.map(({id, row, col}) => {
+
+    if(!isPreset) {
+        findActiveHexagons(e);
+        resendCommand = setInterval(findActiveHexagons, 500, e);
```
seems like this interval will be easy enough to make configurable on the presets page

trueb2 commented on src/Utils/Hexagons.svelte:
```diff
@@ -148,64 +171,178 @@ $: drawableHexagons = hexagons.map(({id, row, col}) => {
+        if (presetName == "sweep") {
+            if (deltaX > 0 && Math.abs(deltaY) < 75 ){
+                    console.log('LR');
+                    $OP_Mode = parseInt("0x86",16);
```
As a rule of thumb, do not use strings to represent integers, unless handling providing user input. This applies to all places using strings when manipulating integers. It can significantly hinder performance, hide bugs, and add surprising code paths.
```suggestion
                    $OP_Mode = 0x86
```

trueb2 commented on stores/stores.js:
```diff
@@ -26,3 +31,25 @@ export const act_command = derived([devices, activeDevice, block0_31, block32_63
+
+export const preset_display = readable([
+  { name: "flashall", 
+    array: [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]]},
```
These should be programmatic and concise. Since we only have a single hexagon configuration, this works for now, but otherwise, this is a bad pattern.

Another rule of thumb, unnamed magic numbers are unacceptable. There should be an obvious trail back to the reason for every hardcoded number. The fewer there are, the fewer mistakes and the easier it is to make changes in the future.
```suggestion
    array: [[...Array(36).keys()]},
```

We can leave the rest of these for a later revision when we handle stitchable fabrics.

trueb2 reviewed (changes requested):
Primarily, we need to fix the multi-touch command generation. Also, remove the WIP title when you are ready to merge.

rschloen committed:
Merge branch 'user/rschloen/touch-handling' into user/rschloen/style-changes

rschloen committed:
Merge pull request #5 from qsib-cbie/user/rschloen/style-changes

Improving Actuator Visuals

rschloen committed:
Merged array sizing changes

rschloen committed:
Added collapsible container

rschloen committed:
Merge branch 'user/rschloen/touch-handling' of https://github.com/qsib-cbie/epidermal-actuators-ui into user/rschloen/touch-handling

rschloen committed:
Added logic for single pulse block

rschloen committed:
Added allOff button

Sets OP mode to 0 and zeros all blocks for good measure (may not be needed)

Bind Hexagon object can calls export sendCommandBlocks to send command.

rschloen committed:
Removed all parseInt

Switched from string rep to decimal or hex rep where possible

rschloen committed:
Improved building of actuator commands

Previous implementation relied on string representation for building the
command to account for multiple blocks of actuators. Math reworked to
calculate correct actuator block and byte block(represented as index of Array(16)),
and assigned the correct decimal representation.

rschloen committed:
Merge branch 'user/rschloen/touch-handling' of https://github.com/qsib-cbie/epidermal-actuators-ui into user/rschloen/touch-handling

rschloen committed:
Added additional logic for turning on single pulse

Checkbox can be used to turn on/off single pulse mode (0x05)

rschloen committed:
Added cache for activeHexagons

Cache can be checked against to not send additional commands when on the
same actuator button.

Moved AllOff to Hexagons component

rschloen committed:
Merge branch 'user/rschloen/touch-handling' of https://github.com/qsib-cbie/epidermal-actuators-ui into user/rschloen/touch-handling

rschloen committed:
Cleaned up collapsable logic/style

rschloen committed:
Made collapsable css global

rschloen committed:
Improved tracking of touch events/points

Renamed tpCache to hexCache and used tpCache to track currect touch
points based on events. When a touch ends, tries to remove it from
tpCache and if nothing left in tpCache, throws error to cleanup touchend.

try catch used to distinguish between mouse and touch events.

Switched to using tpCache in findActiveActuators instead of using touches
from events. Event is still passed since its needed for mouse events.

Cleaned up dead code and logs

rschloen committed:
Made minor changes to timings

rschloen committed:
Linked overlaid arrays to sibling full size arrays

The small array on the option buttons is linked to the full size array for that
option so the same actuator button will light up on both, and both will
have the same layout and rotation.

Also added new background images for the options.

rschloen committed:
Updated act_command to match new protocol

names, values, and calculations for act_command fields were updated to
match the new firmware protocol.

Several fields were renamed to avoid confusion. Ex. OP_mode -> command
since OP_mode now refers to different bytes in the command.

A timing block field was removed to match new firmware. And math was changed
since each timing block now consists of 3 bytes.

rschloen committed:
Added Set Timing button

Set Timing button sets the cmd_op, which is used in the back end, to
set only the timing blocks without sending actuators.

Made cmd_op writable in the store.

rschloen committed:
Fixed setting cmd_op

Used temp to reset $cmd_op when setting timing

Added comment to explain timing

rschloen committed:
Reduced number of links to two

Instead of 4 links to different pages, reduced down to 2: configuration
and Automatic(probably be renamed).

Removed option to hide timing sliders, all slider will always be visible

Removed first column on Automatic page (message, devices)

rschloen committed:
Added button to turn on/off rf field

onclick sends command with state(on or off).

rschloen committed:
Fixed hexCache and single pulse conditions

Fixed condition on resetting hexCache for interval to include both
single pulse options.

rschloen committed:
Removed unneeded display of variable

rschloen committed:
Added variable for act_cnt8 in store

For future support for larger devices.

rschloen committed:
Merge pull request #6 from qsib-cbie/user/rschloen/protocol_update

Updated act_command to match new protocol

rschloen committed:
Added icons to nav buttons

rschloen committed:
Added Icons to manual page components

Removed extra features that hid options for the timing configuration.

Removed unneed css.

rschloen committed:
Refactored link between option and main container

Added images for the different body location options.

Made minor changes to style of scrollbox and option buttons

Switched from using Router to link options and main hexagon to using if
statements and click events. This reduces amount of code since each link
 doesn't need to be specified.

Made one central container for main hexagon array that is conditionally
bound to hexagons on the last button pressed. That means binding is only
active for one button at a time.

rschloen committed:
Removed unneeded code

Not using draggable for svelte moveable since not implementing stiching.

rschloen committed:
Added presets to virtual touch page

A preset is turned on by pressing its toggleable button on the right.
Blue border means its on and a click or swipe on the actuators will
trigger it.

rschloen committed:
Moved onclick inline function for options

Made new local function to handle clicking on an option since it needs
to do more things.

rschloen committed:
Fixed layout for virtual touch page

Virtual touch's layout was set so there is no scrolling and options and
main view take up as much space avaible on differnt screen sizes.

Nav tabs moved to bottom.

Removed some dead code from hexagons.

rschloen committed:
Created Status component

Replaces Messages.
Put in fixed location to stay consistent on all pages.

Power symbol represents Sever connection. Green=connected, red=not
connected

Wifi symbol represents success/failure of communicating with device.
Errors could include failure to configure antenna, or failure in sending
actuator commands.

rschloen committed:
Made arrays for options to loop through

Options buttons are not generated by iterating through array of objects.
Object's properties are used to set the style and id of button, for both
location and preset options.
This simplifies adding new options.

rschloen committed:
Changed how style of main view is changed

When option is clicked, sets main_obj to the object with that options
settings and uses those settings to style the main container.

Added div with relative position around the hexagons so they could be
positioned better over background. Could be replaced with draggable feature
in future

Made position of status component dependent on window size.

rschloen committed:
Made array size dependent on window size

In virtual touch page, made array size of main hexagon array dependent
on window size.

Changed position of hexagon arrays in the option buttons since top %
property wasn't working properly.

Removed extra container from option button to simplify things.

rschloen committed:
Added device property to hexagons

Hexagons will now have a dev property, along with id, row, col, which sets
the activeDevice number for the actuator command.
This lets you build larger arrays using the stichables(groups of 7) with
id numbers 0-7, and using dev property to distinguish the different
devices.

This limits the user to hardcoded arrangments of stichables.

rschloen committed:
Added Manual option to presets

Manual button is outlined whenever no other preset is used.
Used to be implicit, now manual control is explicity indicated.

rschloen committed:
Fixed size of hand and shoulder image

Cropped extra whitespace in images to center them.

rschloen committed:
Merge branch 'user/rschloen/touch-handling' into user/rschloen/page-redesign

rschloen committed:
Added new arrayType

Added hand array type, which consists of 3 stichables.
Added better logic for switching arrayTypes to handle more arrays types.

Reduced size of hexagon arrays.

rschloen committed:
Fixed hex array positioning/sizeing

Added transform:translate(-50%,-50%) to effectively set the origin of the
hexagons to the center instead of top left corner.
This allowed the use of more intuitive percentages for positioning the
arrays in the option buttons.
Also helps account for differences in size, since the buttons and arrays
may be different sizes on different screens so 30% with the origin in the
top left corner would be different.
Seems to also let you use the same values for the main container and
the options.

Set app div to center everything so columns line up better.
Adjusted position of status indicator.
Fixed variable name in store.

rschloen committed:
Merge pull request #7 from qsib-cbie/user/rschloen/page-redesign

Finalize Page Design

rschloen committed:
Removed unneeded code/changes

Moved center from App to automatic since manual did not need it.

Removed unused properties from location_options

rschloen committed:
Made endpoint reactive

Setting hostname, port, route in configuration will change the endpoint.
Allows for changes to endpoint without restarting.

Misc. spacing fix in manual.

rschloen committed:
Made new hexagon layout for config testing

Using stichable designed, remapped hexagons to match
stichables shape in corner of 36 actuator device.

rschloen committed:
Added prop to hexagons to prevent sending of commands

is_active prop is set to false for the arrays over the option buttons to
prevent them from sending any extra commands when pressed.

Added "test" option to hexagon layouts for configuration page test array
so it matches the same shape on the 36 array hexagon.

rschloen committed:
Updated preset commands

Presets have different command bytes in new firmware, so the commands
were updated.

rschloen committed:
Added individual timing vars to store

Adding the individual timing variables (lfperiod, lfdutycycle,...) to
store allows them to be used without conversion in the hexagons component
for things like sleeping for the right time interval during presets.

Also fixes bug where the values would reset to the initial values
whenever you came back to the configuration page.

rschloen committed:
Fixed bug for tracking touch point location

Touch point location is calculated relative to the bounding box of the
target container. The box was previously set onMount to prevent the
target box being changed when moving over text (id#) in the hexagon
buttons.
This caused problems since the target is moved in virutal touch
page but the box is never updated. Since there is no longer text in the
hexagons, assignment of bounding box is done on touch start.

rschloen committed:
Fixed logic related to single pulse mode

rschloen committed:
Updated preset commands for new firmware

Made visual preset deisplay on ui use same timing as device.

rschloen committed:
Added new hexagon layouts

New hexagon layouts for different body locations were hardcoded.
Added rotation property to location_options to rotate hexagon array.

Fixed bug causing overlay array to be live. (variable name conflict)

rschloen committed:
Added preset option images

Added images to buttons for selected presets. Seperated Manual (Live)
button from rest of presets.
Fixed Status width so it no longer overlapped and interfered with
preset button.
Added another line size to arrays based on size of hexagons.

rschloen committed:
Fixed back image and final style tweaks

trueb2 reviewed (approved):
Great job this semester!
